src/rust/engine/src/core.rs

```diff
@@ -351,23 +353,86 @@ impl Failure {
       }
     }
   }
-}
 
-impl Failure {
+  pub fn join_tracebacks(self, new_failure: Self) -> Failure {
+    match (self, new_failure) {
+      (
+        Failure::Throw {
+          val: _old_val,
+          python_traceback: old_python_traceback,
+          engine_traceback: old_engine_traceback,
+        },
+        Failure::Throw {
+          val: new_val,
+          python_traceback: new_python_traceback,
+          engine_traceback: new_engine_traceback,
+        },
+      ) => {
+        let joined_python_tracebacks = [
+          old_python_traceback,
+          "\nWhile handling that, another exception was raised:\n".to_string(),
+          new_python_traceback,
+        ]
+        .join("\n");
+        let joined_engine_tracebacks = old_engine_traceback
+          .into_iter()
+          .chain(new_engine_traceback.into_iter())
+          .collect::<Vec<_>>();
+        Failure::Throw {
+          val: new_val,
+          python_traceback: joined_python_tracebacks,
+          engine_traceback: joined_engine_tracebacks,
+        }
+      }
+      _ => Failure::Invalidated,
+    }
+  }
+
+  pub fn as_py_err(&self) -> Result<&Value, Self> {
+    match &self {
+      &Failure::Invalidated => Err(Failure::Invalidated),
+      &Failure::Throw { val, .. } => Ok(val),
+    }
+  }
+
+  pub fn into_py_err(self) -> Result<Value, Self> {
+    match self {
+      Failure::Invalidated => Err(Failure::Invalidated),
+      Failure::Throw { val, .. } => Ok(val),
+    }
+  }
+
+  pub fn from_py_err_value(val: Value) -> Failure {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+    let obj = val.consume_into_py_object(py);
+    let py_err = PyErr::from_instance(py, obj);
+    Self::from_py_err_with_gil(py, py_err)
+  }
   pub fn from_py_err(py_err: PyErr) -> Failure {
     let gil = Python::acquire_gil();
     let py = gil.python();
     Failure::from_py_err_with_gil(py, py_err)
   }
   pub fn from_py_err_with_gil(py: Python, mut py_err: PyErr) -> Failure {
-    let val = Value::from(py_err.instance(py));
-    let python_traceback = if let Some(tb) = py_err.ptraceback.as_ref() {
+    let err_object = py_err.instance(py);
+    // If executing generator_send() or generator_throw() raises an exception, we add a
+    // ._formatted_stack attribute which is used to persist the stacktrace across @rule methods.
+    let python_traceback = if err_object.hasattr(py, "_formatted_stack").unwrap_or(false) {
+      let locals = PyDict::new(py);
+      locals.set_item(py, "err_object", &err_object).unwrap();
+      py.eval("''.join(err_object._formatted_stack)", None, Some(&locals))
+        .unwrap()
+        .extract::<String>(py)
+        .unwrap()
+    } else if let Some(tb) = py_err.ptraceback.as_ref() {
+      // Otherwise, we rely on the traceback from the cpython crate.
```

Comment:
The reason we need the `._formatted_stack` part at all is because `ptraceback` isn't getting set when we re-raise an exception. I think that's probably fine, and I think there's still value to keeping that branch in case we start using `Failure` elsewhere in places that don't get `._formatted_stack` set. But I haven't investigated why the `ptraceback` is only set sometimes. cc @gshuflin